src/TruePBR.cpp

```diff
@@ -0,0 +1,1505 @@
+				if (singleton->IsPBRTextureSet(land->loadedData->defQuadTextures[quadIndex]->textureSet)) {
+					isPbr = true;
+					break;
+				}
+			}
+			for (uint32_t textureIndex = 0; textureIndex < 6; ++textureIndex) {
+				if (land->loadedData->quadTextures[quadIndex][textureIndex] != nullptr) {
+					if (singleton->IsPBRTextureSet(land->loadedData->quadTextures[quadIndex][textureIndex]->textureSet)) {
+						isPbr = true;
+						break;
+					}
+				}
+			}
+		}
+	}
+
+	if (!isPbr) {
+		return ptr_TESObjectLAND_SetupMaterial(land);
+	}
+
+	static const auto settings = RE::INISettingCollection::GetSingleton();
+	static const bool bEnableLandFade = settings->GetSetting("bEnableLandFade:Display");
+	static const bool bDrawLandShadows = settings->GetSetting("bDrawLandShadows:Display");
+	static const bool bLandSpecular = settings->GetSetting("bLandSpecular:Landscape");
+
+	if (land->loadedData != nullptr && land->loadedData->mesh[0] != nullptr) {
+		land->data.flags.set(static_cast<RE::OBJ_LAND::Flag>(8));
+		for (uint32_t quadIndex = 0; quadIndex < 4; ++quadIndex) {
+			auto shaderProperty = static_cast<RE::BSLightingShaderProperty*>(RE::MemoryManager::GetSingleton()->Allocate(sizeof(RE::BSLightingShaderProperty), 0, false));
+			shaderProperty->Ctor();
+
+			{
+				BSLightingShaderMaterialPBRLandscape srcMaterial;
+				shaderProperty->LinkMaterial(&srcMaterial, true);
+			}
+
+			auto material = static_cast<BSLightingShaderMaterialPBRLandscape*>(shaderProperty->material);
+			const auto& stateData = RE::BSGraphics::State::GetSingleton()->GetRuntimeData();
+
+			material->diffuseTexture = stateData.defaultTextureBlack;
+			material->normalTexture = stateData.defaultTextureNormalMap;
+			material->landscapeDisplacementTextures[0] = stateData.defaultTextureBlack;
+			material->landscapeRMAOSTextures[0] = stateData.defaultTextureWhite;
+			for (uint32_t textureIndex = 0; textureIndex < BSLightingShaderMaterialPBRLandscape::NumTiles - 1; ++textureIndex) {
+				material->landscapeBaseColorTextures[textureIndex] = stateData.defaultTextureBlack;
+				material->landscapeNormalTextures[textureIndex] = stateData.defaultTextureNormalMap;
+				material->landscapeDisplacementTextures[textureIndex + 1] = stateData.defaultTextureBlack;
+				material->landscapeRMAOSTextures[textureIndex + 1] = stateData.defaultTextureWhite;
+			}
+
+			if (auto defTexture = land->loadedData->defQuadTextures[quadIndex]) {
+				SetupLandscapeTexture(*material, *defTexture, 0);
+			}
+			for (uint32_t textureIndex = 0; textureIndex < 6; ++textureIndex) {
+				if (auto landTexture = land->loadedData->quadTextures[quadIndex][textureIndex]) {
+					SetupLandscapeTexture(*material, *landTexture, textureIndex + 1);
+				}
+			}
+
+			if (bEnableLandFade) {
+				shaderProperty->unk108 = false;
+			}
+
+			bool noLODLandBlend = false;
+			auto tes = RE::TES::GetSingleton();
+			if (tes->worldSpace != nullptr) {
+				if (auto terrainManager = tes->worldSpace->GetTerrainManager()) {
```

Comment:
`tes->worldSpace` should be `tes->GetRuntimeData2().worldSpace` for AE/SE

Reply:
Done